### PR TITLE
Optionally use personal access token

### PR DESCRIPTION
**Pull request by WillAbides**
This implements #44.

There are a few things left to do
- [x] Show field errors on the user-preferences form
- [x] Work out a way to submit user-preferences without changing the token.  That isn't that important now, but we will probably have more preferences in the future.
- [x] Don't show personal access token on the user-preferences form if that is disabled
- [x] Add text to the user-preferences form explaining what a personal access token is and how to create one.
- [x] Add an environment variable to enable the use of personal access tokens.

I want to call out the last one especially.  It isn't mentioned in #44, but I think we should have this feature disabled by default.  Because these tokens are equivalent to GitHub passwords and we are storing them in plain text, it wouldn't be responsible to store them in a shared environment.

### CODE REVIEW

**Comment by andrew** [comment]
> Add an environment variable to enable the use of personal access tokens.

Agreed, making that opt in is a great idea 👌

**Comment by andrew** [comment]
@WillAbides we should also add some documentation around this feature in the readme as it's more aimed at people running their own instance than for the hosted version.

File: test/controllers/users_controller_test.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 require 'test_helper'
+require 'pp'
```
Comment (andrew) [inline test/controllers/users_controller_test.rb]: we can remove this right?

Comment (WillAbides) [inline test/controllers/users_controller_test.rb]: Oops.  I'll remove it in my next commit.

**Comment by WillAbides** [comment]
I've gone through my todo list and added to readme.  Removing [WIP] from the title, and this is for review.

**Comment by andrew** [comment]
Looking good @WillAbides, just need a `ensure_logged_in` before_action on the users controller to make sure the user is logged in

**Comment by tarebyte** [comment]
> just need a `ensure_logged_in` before_action on the users controller to make sure the user is logged in

What about adding the `before_action` to the `ApplicationController` and then skipping it in the `SessionsController`? That way we default to being logged in.

**Comment by andrew** [comment]
@tarebyte currently the notifications controller also renders the homepage 🙈

**Comment by WillAbides** [comment]
I ended up using the `before_action :authenticate_user!` that is inherited from ApplicationController and adding `ensure_correct_user` to make sure you can't update a different user.

**Comment by andrew** [comment]
LGTM, any objections @tarebyte?

**Comment by WillAbides** [comment]
@andrew: You said redirect to /login before my last push.  I guess you are ok with it redirecting to / instead?

**Comment by andrew** [comment]
@WillAbides oh yeah that's fine, the session controller would forget where they came from anyway!

File: app/models/user.rb [hunk 2]
```diff
@@ -16,7 +23,7 @@ def assign_from_auth_hash(auth_hash)
     github_attributes = {
       github_id: auth_hash['uid'],
       github_login: auth_hash['info']['nickname'],
-      access_token: auth_hash.dig('credentials', 'token')
+      access_token: auth_hash.dig('credentials', 'token'),
```
Comment (tarebyte) [inline app/models/user.rb]: I know some people prefer to have to have a trailing commas to make the diff cleaner, but my inner rubocop is sad.

What do you think about undoing this change?

File: config/initializers/octobox_config.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+module Octobox
```
Comment (tarebyte) [inline config/initializers/octobox_config.rb]: Does this need to load before any other configs?

If so, might be helpful to prefix this with a `02`

File: app/models/user.rb (hunk 2 above)
Comment (WillAbides) [inline app/models/user.rb]: It's gone.  This was purely unintentional.

File: config/initializers/octobox_config.rb (hunk 3 above)
Comment (WillAbides) [inline config/initializers/octobox_config.rb]: The order shouldn't matter here.

File: config/routes.rb [hunk 4]
```diff
@@ -20,4 +20,7 @@
       get :star
     end
   end
+
+  get '/user_preferences', to: 'users#edit'
```
Comment (tarebyte) [inline config/routes.rb]: Not a huge fan of `user_preferences` as a url. What do you think about changing it to `settings`?

**Comment by tarebyte** [comment]
Left a few comments, looks fine to me @andrew

File: config/routes.rb (hunk 4 above)
Comment (WillAbides) [inline config/routes.rb]: Settings is fine with me.  My main reason for changing it from /edit was to avoid ambiguity about what was being edited.

**Comment by andrew** [comment]
Cool! We can do https://github.com/octobox/octobox/issues/175 nicely once this is merged

**Comment by WillAbides** [comment]
I was already thinking about doing #175 next.

**Comment by andrew** [comment]
:shipit:

**Comment by bronislav** [comment]
@andrew Is this feature available at octobox.io ?

**Comment by andrew** [comment]
@bronislav no it's not at the moment, I didn't feel comfortable storing people's personal access tokens plaintext the heroku database.

**Comment by bronislav** [comment]
@andrew Thank you for clarification

**Comment by andrew** [comment]
@bronislav I've opened #212 to get those features documented in the readme

**Pull request merged**